### ISSUE TITLE
getRedirectResult() always returns null - Unable to retrieve idToken after signInWithRedirect()

### ISSUE THREAD

**Issue by falk-stefan** [angular/angularfire#3335]
### Version info

**Angular:** 15

**Firebase:** 9

**AngularFire:** 7.5.0

### How to reproduce these conditions

**Steps to set up and reproduce**

I'm simply calling `signInWithRedirect()` and install an `APP_INITIALIZER` provider to make sure the `idToken` is loaded as the app starts.

However, the `idToken` (as well as the `credentials`) are always `null`/`undefined` and I have no idea why. Could be related to (https://github.com/angular/angularfire/issues/3208). 

```typescript
public signInWithGoogle() {
  return signInWithRedirect(this.auth, new GoogleAuthProvider());
}
```

```typescript
export const AUTH_ID_TOKEN_INITIALIZER_PROVIDER: Provider = {
  provide: APP_INITIALIZER,
  deps: [NGXLogger, AuthService],
  multi: true,
  useFactory: (logger: NGXLogger, authService: AuthService) => {
    return async () => {
      logger.info("Initializing App: Waiting for (API) idToken");
      const credentials = await getRedirectResult(authService.auth);
      console.log(credentials);  // Is already null
      const idToken = await credentials?.user.getIdToken();
      console.log(idToken);
      if (idToken) {
        return idToken;
      }
      return firstValueFrom(authService.idToken$);
    }
  },
};
```  

I tried a different approach after reading that we have to wait for `onAuthStateChanged()` to fire and only after that we're supposed to fetch the `idToken`. The implementation looked like this (with the same result however):

```typescript
async init() {
  this.logger.debug("Checking for token from redirect");
  const credentials = await getRedirectResult(this.auth);
  this.idToken = await credentials?.user.getIdToken();

  await this.auth.onAuthStateChanged(
    async () => {
      const credentials = await getRedirectResult(this.auth);
      this.idToken = await credentials?.user.getIdToken();
      console.log("inner");
      console.log(this.idToken);
    }
  );
  console.log("outer");
  console.log(this.idToken);
}
``` 

### Expected behavior

Receive a valid `idToken` after beind redirected.

### Actual behavior

Credentials & token are `null` / `undefined`.

**Comment by falk-stefan** [comment]
@rgant No, I'm not using incognito here. It's neither working on mobile nor on desktop.

I noticed that the redirect works if I am using the emulators - for whatever reason - but emulators are [breaking the Spring Boot authentication chain](https://stackoverflow.com/questions/76005413/how-to-work-with-spring-boot-angular-under-firebase-authentication). 

It's painfully annoying..

**Comment by rgant** [comment]
I just implemented this myself, and best I can tell it is working just fine.

My advice would be to join the [Angular Community Discord](https://discord.gg/angular) and ask there. 

What I am seeing in your code above is that you are _always_ checking `getRedirectResult` which will emit an object with `null` credentials and user properties if the `signInWithRedirect` call either wasn't made in the past, or wasn't able to write to storage. My solution for this was to set a flag in storage and then only call `getRedirectResult` if the flag was there. But I was using a specific page for my redirect login.

Note that [Google's own example code](https://github.com/firebase/quickstart-js/blob/7f83300a91ed7f06ee705e186c338eaf06347d18/auth/google-redirect.html#L63) expects you to have to test that the credential property is truthy.

**Comment by falk-stefan** [comment]
@rgant Can you elaborate what your exact setup is and what it is that is working on your end?

`getRedirectResult()`, as stated, only works when I am using the Authentication Emulator in which case I'll get a non-null `UserCredential` which holds a "valid" token. However, this token doesn't fly with my backend as it cannot be verified under https://securetoken.google.com for obvious reasons.

This is annoying because I cannot just use local test-accounts on my client and test client/server together. 

The core I posted above looks "weird" and lacks some checking because it is me jolting everything back and forth hoping it'll work somehow.

So.. is `getRedirectResult()` workign for you with and without emulators?

**Comment by falk-stefan** [comment]
@jhades I'm still struggling with this.

**Comment by ohabash** [comment]
also struggling with this. my user is logged 

```
return this.afAuth
			.signInWithPopup(provider)
			.then((userCredentials) => this.loginSuccess(userCredentials));
```

now in another part of the app at need these credentials again. so im trying 

```
const result = await this.afAuth.getRedirectResult();
console.log(`🚀 => AuthInterceptor => result:`, result) // always null
```

but the result is `{user: null}`

**Comment by IliaIdakiev** [comment]
I'm also experiencing the same issue, the `getRedirectResult()` was working when I started developing the project and then all of a sudden it started returning `null`

**Comment by marius-eichenberg** [comment]
@falk-stefan, are you, by any chance, protecting a route with an auth guard?

**Comment by falk-stefan** [comment]
@marius-eichenberg Uh, that's a good points. I can't really tell right now as I am on a different project currently, but there's a possibility an auth-guard interferred here.

**Comment by servefast-cto** [comment]
Same issue with me its becoming annoying always something not working with the emulator

**Comment by Romanovskyi2137** [comment]
so, still no solution right?) actually i have the same problem, BUT, earlier my implementation worked well, but now it`s don`t, i have null after come back to login page after redirect. magic

**Comment by rlunaro** [comment]
I've reproduced this error in a simple angular project. It fails as described: after log in with google, it reads: it seems that this is not working and "null". In other words, a null user is received, plain and simple. One component, one angular application. 

You can check this out here: https://github.com/rlunaro/null-test

@rgant  can check what is going on on this project, I will appreciate that.

**Comment by qw789** [comment]
the same problem.

**Comment by alsami** [comment]
Have the exact same problem, login stopped working when using `signInWithRedirect` without any code change. Switched to `signInWithPopup` for the meanwhile but that is not a solution.

**Comment by rlunaro** [comment]
I've found a solution for the problem. The problem apparently is related with the technology used to pass the credentials and the information between your website and the identity provider (google, facebook, apple, microsoft or whatever). 

For that, Firebase apparently uses a _hidden iframe_. The problem is that modern browsers doesn't allow the communication between hidden iframes if they doesn't belong to authorized domains. 

**In a nutshell:** you cannot test locally the authentication process against google, microsoft or an identity provider using firebase. **You can do that when you deploy your website**, because the authorized domain (the domain your application is hosted _is authorized_ to query a hidden iframe that is used to transfer the information. 

Locally, you can use the firebase auth emulator. First, start the emulator: 

```bash
firebase emulators:start --only auth
```

then, connect your application to the emulator by adding the following line: 

```
connectAuthEmulator( this.auth, "http://localhost:9099" );
```

Where 'this.aut' must be the firebase auth object.

In my case that worked and when I deployed the application -prior removing the former line-, it worked like a charm. In the correct domain, the application don't have any problems.

**Comment by Pedro-Marques-Santos** [comment]
I spent days struggling with signInWithRedirect in Firebase. When I switch from popup to redirect, the result on the main page is always null. The Google login redirects just fine, but getRedirectResult() never returns the user.

Here’s what actually fixed it:

- I ran the project using ngrok to get a real HTTPS domain https://my-ngrok.ngrok-free.app
- Set that domain in firebaseConfig as the authDomain
- Added it to the authorized domains in Firebase and as a redirect URI and JavaScript origin in the Google Cloud OAuth 2.0 client
- Used NGINX locally to proxy the /__/auth route to Firebase Hosting. Without this, it would return 404 and the login wouldn't complete.


NGINX: 

```
server {
    listen 80;
    server_name https://my-ngrok.ngrok-free.app;

    location / {
        proxy_pass http://111.111.1.1111:3000;     // local IP
        proxy_http_version 1.1;
        proxy_set_header Upgrade $http_upgrade;
        proxy_set_header Connection 'upgrade';
        proxy_set_header Host $host;
        proxy_cache_bypass $http_upgrade;
    }

    location /__/auth {
        proxy_pass https://firebase.firebaseapp.com;  // your firebase redirect authDomain
        proxy_ssl_server_name on;
    }
}
```


With all that in place, I was finally able to test Google login with redirect locally

**Comment by armando-navarro** [comment]
Thanks for the report and for the workarounds shared in the thread (the reverse-proxy and ngrok recipes are genuinely useful).

My reading is that this is not something AngularFire can fix in code: `signInWithRedirect` and `getRedirectResult` in `@angular/fire/auth` are thin re-exports of the firebase-js-sdk functions, and the null result comes from the SDK plus browser storage policy. Completing the redirect involves reading the result back through a cross-origin request to your `authDomain` (usually `<project>.firebaseapp.com`), and modern browsers (Chrome's third-party storage partitioning, Safari ITP) block that cross-site storage access when your app origin differs from the `authDomain`, so the result cannot be retrieved and comes back null.

A few things that tend to resolve it:

- Serve the app same-origin with the auth handler by reverse-proxying the `/__/auth/` path to your Firebase Hosting `authDomain` (the NGINX approach @Pedro-Marques-Santos posted), or use a custom same-origin `authDomain`.
- Where a redirect is not required, `signInWithPopup` avoids the cross-origin storage dependency.
- For local development, use the Auth emulator, or a real HTTPS tunnel domain added to your Firebase authorized domains and Google OAuth origins.
- In code, `getRedirectResult` resolves the pending redirect only once, and resolves to `null` when there was no pending redirect in this context. So if a route guard or an earlier call already read it, a later call returns null. Read it in exactly one place, and react to `authState`/`onAuthStateChanged` for the resolved user rather than relying on `getRedirectResult` alone.

Firebase has a "Best practices for signInWithRedirect" guide that covers the third-party-cookie situation in more depth. If you are hitting this on a same-origin production setup where none of the above applies, please say so and share the setup, since that would point somewhere I have not.
